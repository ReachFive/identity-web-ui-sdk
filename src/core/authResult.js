import { parseJwtTokenPayload } from '../helpers/jwt'
import { logError } from '../helpers/logger'

export function enrichAuthResult(response) {
    if (response.idToken) {
        try {
            const idTokenPayload = parseJwtTokenPayload(response.idToken);

            return {
                ...response,
                idTokenPayload
            };
        } catch (e) {
            logError('id token parsing error: ' + e);
        }
    }

    return response;
}

export function isAuthResult(value) {
    return value && (value.accessToken || value.idToken);
}

export function isErrorResponse(value) {
    return value && value.error;
}
