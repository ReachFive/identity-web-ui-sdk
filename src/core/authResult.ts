import { AuthResult, ErrorResponse } from '@reachfive/identity-core';

import { parseJwtTokenPayload } from '../helpers/jwt';
import { logError } from '../helpers/logger';

export function enrichAuthResult(response: AuthResult) {
    if (response.idToken) {
        try {
            const idTokenPayload = parseJwtTokenPayload(response.idToken);

            return {
                ...response,
                idTokenPayload,
            };
        } catch (e) {
            logError('id token parsing error: ' + e);
        }
    }

    return response;
}

export function isAuthResult(value: unknown): value is AuthResult {
    return (
        value !== null &&
        typeof value === 'object' &&
        ('accessToken' in value || 'idToken' in value)
    );
}

export function isErrorResponse(value: unknown): value is ErrorResponse {
    return value !== null && typeof value === 'object' && 'error' in value;
}
