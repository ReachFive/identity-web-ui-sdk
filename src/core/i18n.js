import isPlainObject from 'lodash-es/isPlainObject';

export default function resolveI18n(defaultMessages, messages = {}) {
    const mergedMessages = {
        ...defaultMessages,
        ...flattenObject(messages)
    };

    return (key, params) => {
        const template = mergedMessages[key] || key;

        return params
            ? Object.keys(params).reduce((acc, param) => acc.replace(`{${param}}`, params[param]), template)
            : template;
    };
}

function flattenObject(object, prefix = []) {
    return isPlainObject(object)
        ? Object.keys(object).reduce((acc, key) => ({
            ...acc,
            ...flattenObject(object[key], [...prefix, key])
        }), {})
        : ({ [prefix.join('.')]: object });
}
