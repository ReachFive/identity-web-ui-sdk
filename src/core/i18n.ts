export type I18nMessages = Record<string, string>

export type I18nNestedMessages = Record<string, string | I18nMessages>

export type I18nMessageParams = Record<string, unknown>

export type I18nResolver = (key: string, params?: I18nMessageParams, fallback?: (params?: I18nMessageParams) => string) => string

export function resolveI18n<T extends (Record<string, T> | string)>(defaultMessages: I18nMessages = {}, messages: I18nNestedMessages = {}): I18nResolver {
    const mergedMessages: I18nMessages = {
        ...defaultMessages,
        ...flattenObject(messages)
    };

    return (key: string, params?: I18nMessageParams, fallback?: (params?: I18nMessageParams) => string) => {
        const template = mergedMessages[key] ?? fallback?.(params) ?? key;

        return params
            ? Object.keys(params).reduce((acc, param) => acc.replace(`{${param}}`, params[param] as string), template)
            : template;
    };
}

export default resolveI18n

function flattenObject(object: I18nNestedMessages | string, prefix: string[] = []): Record<string, string> {
    return typeof object === "object" && object !== null
        ? Object.keys(object).reduce((acc, key) => ({
            ...acc,
            ...flattenObject(object[key], [...prefix, key])
        }), {})
        : ({ [prefix.join('.')]: object });
}
