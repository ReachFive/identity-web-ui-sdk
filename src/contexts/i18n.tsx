import React, { ComponentType, PropsWithChildren, useMemo } from 'react';

import { I18nMessages, I18nNestedMessages, I18nResolver, resolveI18n } from '../core/i18n';

export interface Props {
  defaultMessages?: I18nMessages
  messages?: I18nNestedMessages
}

export const I18nContext = React.createContext<I18nResolver | undefined>(undefined);

export function useI18n(): I18nResolver {
  const context = React.useContext(I18nContext);
  if (!context) {
    throw new Error('No I18nContext provided');
  }

  return context;
}

export interface I18nProps {
  i18n: I18nResolver
}

export type WithI18n<P> = P & I18nProps

export function withI18n<T extends I18nProps = I18nProps>(WrappedComponent: ComponentType<T>) {
  const displayName = WrappedComponent.displayName || WrappedComponent.name || "Component";

  const ComponentWithI18n = (props: Omit<T, keyof I18nProps>) => {
    const i18n = useI18n();
    return <WrappedComponent {...({ i18n })} {...(props as T)} />;
  };

  ComponentWithI18n.displayName = `withI18n(${displayName})`;

  return ComponentWithI18n;
}

export function I18nProvider({ children, defaultMessages, messages }: PropsWithChildren<Props>): JSX.Element | null {
  const resolver = useMemo(() => resolveI18n(defaultMessages, messages), [defaultMessages, messages])
  return <I18nContext.Provider value={resolver}>{children}</I18nContext.Provider>
}
