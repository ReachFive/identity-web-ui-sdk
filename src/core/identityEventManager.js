import { enrichAuthResult } from './authResult'
import EventManager from '../helpers/eventManager'

export function createEventManager() {
    const eventManager = new EventManager();

    return {
        on(eventName, listener) {
            eventManager.on(eventName, listener);
        },

        off(eventName, listener) {
            eventManager.off(eventName, listener)
        },

        fireEvent(eventName, data) {
            if (eventName === 'authenticated') {
                eventManager.fire(eventName, enrichAuthResult(data));
            } else {
                eventManager.fire(eventName, data);
            }
        }
    }
}
