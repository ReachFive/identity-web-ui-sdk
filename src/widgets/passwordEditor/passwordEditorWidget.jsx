import React from 'react';

import omit from 'lodash-es/omit';
import compact from 'lodash-es/compact';
import compose from 'recompose/compose';

import { Validator } from '../../core/validation';

import { createWidget } from '../../components/widget/widget';
import { createForm } from '../../components/form/formComponent';
import { passwordField } from '../../components/form/fields/passwordField'
import { simplePasswordField } from '../../components/form/fields/simplePasswordField';
import { withConfig, withI18n } from '../../components/widget/widgetContext'

export const PasswordEditorForm = compose(withI18n, withConfig)(createForm({
    prefix: 'r5-password-editor-',
    fields({ promptOldPassword, canShowPassword, config }) {
        return compact([
            promptOldPassword && simplePasswordField({
                key: 'old_password',
                label: 'oldPassword'
            }),
            passwordField({
                label: 'newPassword',
                autoComplete: 'new-password',
                canShowPassword
            }, config),
            simplePasswordField({
                key: 'password_confirmation',
                label: 'passwordConfirmation',
                autoComplete: 'new-password',
                validator: new Validator({
                    rule: (value, ctx) => value === ctx.fields.password.value,
                    hint: 'passwordMatch'
                })
            })
        ]);
    },
    resetAfterSuccess: true,
    resetAfterError: true
}));

class PasswordEditor extends React.Component {
    static defaultProps = {
        onSuccess: () => { },
        onError: () => { }
    };

    handleSubmit = data => {
        return this.props.apiClient.updatePassword({
            ...omit(data, ['passwordConfirmation']),
            ...this.props.authentication
        });
    };

    render() {
        const {
            promptOldPassword,
            showLabels,
            onSuccess,
            onError,
            canShowPassword,
        } = this.props;

        return <PasswordEditorForm
            handler={this.handleSubmit}
            supportMultipleSubmits={promptOldPassword}
            promptOldPassword={promptOldPassword}
            canShowPassword={canShowPassword}
            showLabels={showLabels}
            onSuccess={onSuccess}
            onError={onError} />
    }
}

const resolveAuthentication = (accessToken, userId) => {
    if (accessToken) {
        return { authentication: { accessToken } };
    } else if (userId) {
        return { authentication: { userId } };
    } else {
        return {};
    }
};

export default createWidget({
    component: PasswordEditor,
    prepare: options => ({
        promptOldPassword: false,
        ...options,
        ...resolveAuthentication(options.accessToken, options.userId)
    })
});
