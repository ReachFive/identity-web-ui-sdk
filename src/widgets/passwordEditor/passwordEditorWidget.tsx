import React from 'react';

import { Validator } from '../../core/validation';

import passwordField from '../../components/form/fields/passwordField';
import simplePasswordField from '../../components/form/fields/simplePasswordField';
import { createForm, FormContext } from '../../components/form/formComponent';
import { createWidget } from '../../components/widget/widget';

import { useReachfive } from '../../contexts/reachfive';

import type { OnError, OnSuccess } from '../../types';

export type PasswordEditorFormData = {
    password: string;
    passwordConfirmation: string;
    oldPassword: string;
};

interface PasswordEditorFormProps {
    /**
     * Ask for the old password before entering a new one.
     * @default false
     */
    promptOldPassword?: boolean;
    /**
     * Whether or not to provide the display password in clear text option.
     * @default false
     */
    canShowPassword?: boolean;
}

export const PasswordEditorForm = createForm<PasswordEditorFormData, PasswordEditorFormProps>({
    prefix: 'r5-password-editor-',
    fields({ promptOldPassword = false, canShowPassword = false, config }) {
        return [
            ...(promptOldPassword
                ? [
                      simplePasswordField({
                          key: 'old_password',
                          label: 'oldPassword',
                      }),
                  ]
                : []),
            passwordField(
                {
                    label: 'newPassword',
                    autoComplete: 'new-password',
                    canShowPassword,
                },
                config
            ),
            simplePasswordField({
                key: 'password_confirmation',
                label: 'passwordConfirmation',
                autoComplete: 'new-password',
                validator: new Validator<string, FormContext<PasswordEditorFormData>>({
                    rule: (value, ctx) => {
                        return value === ctx.fields.password;
                    },
                    hint: 'passwordMatch',
                }),
            }),
        ];
    },
    resetAfterSuccess: true,
    resetAfterError: true,
});

export interface PasswordEditorProps extends PasswordEditorFormProps {
    /**
     * The authorization credential JSON Web Token (JWT) used to access the ReachFive API, less than five minutes old.
     */
    accessToken?: string;
    /**
     * @toto missing description
     */
    authentication?: Authentication;
    /**
     * @toto missing description
     */
    userId?: string;
    /**
     * Whether the form fields' labels are displayed on the form view.
     * @default false
     */
    showLabels?: boolean;
    /**
     * Callback function called when the request has succeed.
     */
    onSuccess?: OnSuccess;
    /**
     * Callback function called when the request has failed.
     */
    onError?: OnError;
}

const PasswordEditor = ({
    authentication,
    canShowPassword = false,
    promptOldPassword = false,
    showLabels = false,
    onSuccess = () => {},
    onError = () => {},
}: PasswordEditorProps) => {
    const coreClient = useReachfive();

    const handleSubmit = ({ password, oldPassword }: PasswordEditorFormData) => {
        return coreClient.updatePassword({
            password,
            ...(promptOldPassword ? { oldPassword } : {}),
            ...authentication,
        });
    };

    return (
        <PasswordEditorForm
            handler={handleSubmit}
            supportMultipleSubmits={promptOldPassword}
            promptOldPassword={promptOldPassword}
            canShowPassword={canShowPassword}
            showLabels={showLabels}
            onSuccess={() => onSuccess({ name: 'password_changed' })}
            onError={onError}
        />
    );
};

type Authentication = { accessToken: string } | { userId: string };

const resolveAuthentication = (
    accessToken?: string,
    userId?: string
): { authentication?: Authentication } => {
    if (accessToken) {
        return { authentication: { accessToken } };
    } else if (userId) {
        return { authentication: { userId } };
    } else {
        return {};
    }
};

export type PasswordEditorWidgetProps = Omit<PasswordEditorProps, 'authentication'>;

export default createWidget<PasswordEditorWidgetProps, PasswordEditorProps>({
    component: PasswordEditor,
    prepare: options => ({
        ...options,
        ...resolveAuthentication(options.accessToken, options.userId),
    }),
});
