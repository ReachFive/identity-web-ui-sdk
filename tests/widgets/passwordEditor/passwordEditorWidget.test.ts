/**
 * @jest-environment jsdom
 */

import { beforeEach, describe, expect, jest, test } from '@jest/globals';
import '@testing-library/jest-dom/jest-globals';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import 'jest-styled-components';

import type { Client, PasswordStrengthScore } from '@reachfive/identity-core';

import { type I18nMessages } from '../../../src/core/i18n';
import type { Config } from '../../../src/types';

import passwordEditorWidget from '../../../src/widgets/passwordEditor/passwordEditorWidget';

const defaultConfig: Config = {
    clientId: 'local',
    domain: 'local.reach5.net',
    sso: false,
    sms: false,
    webAuthn: false,
    language: 'fr',
    pkceEnforced: false,
    isPublic: true,
    socialProviders: ['facebook', 'google'],
    customFields: [],
    resourceBaseUrl: 'http://localhost',
    mfaSmsEnabled: false,
    mfaEmailEnabled: false,
    rbaEnabled: false,
    consentsVersions: {},
    passwordPolicy: {
        minLength: 8,
        minStrength: 2,
        allowUpdateWithAccessTokenOnly: true,
    },
};

const defaultI18n: I18nMessages = {};

const getPasswordStrengthImplementation = (password: string) => {
    let score = 0;
    if (password.match(/[a-z]+/)) score++;
    if (password.match(/[0-9]+/)) score++;
    if (password.match(/[^a-z0-9]+/)) score++;
    if (password.length > 8) score++;
    return Promise.resolve({ score: score as PasswordStrengthScore });
};

describe('Snapshot', () => {
    const generateSnapshot =
        (
            options: Partial<Parameters<typeof passwordEditorWidget>[0]> = {},
            config: Partial<Config> = {}
        ) =>
        async () => {
            // @ts-expect-error partial Client
            const apiClient: Client = {
                getPasswordStrength: jest
                    .fn<Client['getPasswordStrength']>()
                    .mockImplementation(getPasswordStrengthImplementation),
                updatePassword: jest.fn<Client['updatePassword']>().mockResolvedValue(),
            };

            const widget = await passwordEditorWidget(
                { ...options, accessToken: 'azerty' },
                { apiClient, config: { ...defaultConfig, ...config }, defaultI18n }
            );

            await waitFor(async () => {
                const { container } = await render(widget);
                expect(container).toMatchSnapshot();
            });
        };

    describe('password editor', () => {
        test('basic', generateSnapshot({}));
    });
});

describe('DOM testing', () => {
    const getPasswordStrength = jest
        .fn<Client['getPasswordStrength']>()
        .mockImplementation(getPasswordStrengthImplementation);
    const updatePassword = jest.fn<Client['updatePassword']>();

    const onError = jest.fn();
    const onSuccess = jest.fn();

    beforeEach(() => {
        getPasswordStrength.mockClear();
        updatePassword.mockClear();
        onError.mockClear();
        onSuccess.mockClear();
    });

    const generateComponent = async (
        options: Partial<Parameters<typeof passwordEditorWidget>[0]> = {},
        config: Partial<Config> = {}
    ) => {
        // @ts-expect-error partial Client
        const apiClient: Client = {
            getPasswordStrength,
            updatePassword,
        };

        const result = await passwordEditorWidget(
            { onError, onSuccess, ...options, accessToken: 'azerty' },
            { config: { ...defaultConfig, ...config }, apiClient, defaultI18n }
        );

        return waitFor(async () => {
            return render(result);
        });
    };

    describe('passwordEditor', () => {
        test('default', async () => {
            const user = userEvent.setup();

            updatePassword.mockResolvedValue();

            await generateComponent({});

            expect(screen.queryByLabelText('oldPassword')).not.toBeInTheDocument();

            const newPasswordInput = screen.getByLabelText('newPassword');
            expect(newPasswordInput).toBeInTheDocument();

            const passwordConfirmationInput = screen.getByLabelText('passwordConfirmation');
            expect(passwordConfirmationInput).toBeInTheDocument();

            await userEvent.clear(newPasswordInput);
            await userEvent.type(newPasswordInput, 'bob@reach5.co');

            await userEvent.clear(passwordConfirmationInput);
            await userEvent.type(passwordConfirmationInput, 'bob@reach5.co');

            const submitBtn = screen.getByRole('button', { name: 'send' });
            expect(submitBtn).toBeInTheDocument();

            await user.click(submitBtn);

            expect(updatePassword).toBeCalledWith(
                expect.objectContaining({
                    accessToken: 'azerty',
                    password: 'bob@reach5.co',
                })
            );

            expect(onSuccess).toBeCalled();
            expect(onError).not.toBeCalled();
        });

        test('with promptOldPassword = true', async () => {
            const user = userEvent.setup();

            updatePassword.mockResolvedValue();

            await generateComponent({ promptOldPassword: true });

            const oldPasswordInput = screen.getByLabelText('oldPassword');
            expect(oldPasswordInput).toBeInTheDocument();

            const newPasswordInput = screen.getByLabelText('newPassword');
            expect(newPasswordInput).toBeInTheDocument();

            const passwordConfirmationInput = screen.getByLabelText('passwordConfirmation');
            expect(passwordConfirmationInput).toBeInTheDocument();

            await userEvent.clear(oldPasswordInput);
            await userEvent.type(oldPasswordInput, 'My0ld_Pa55w0rD');

            await userEvent.clear(newPasswordInput);
            await userEvent.type(newPasswordInput, 'Wond3rFu11_Pa55w0rD*$');

            await userEvent.clear(passwordConfirmationInput);
            await userEvent.type(passwordConfirmationInput, 'Wr0ng_Pa55w0rD*$'); // don't match newPasswordInput

            const submitBtn = screen.getByRole('button', { name: 'send' });
            expect(submitBtn).toBeInTheDocument();

            await user.click(submitBtn);

            // password confirm don't match
            const formError = screen.queryByTestId('form-error');
            expect(formError).toBeInTheDocument();
            expect(formError).toHaveTextContent('validation.passwordMatch');

            await userEvent.clear(passwordConfirmationInput);
            await userEvent.type(passwordConfirmationInput, 'Wond3rFu11_Pa55w0rD*$'); // match newPasswordInput

            await user.click(submitBtn);

            expect(updatePassword).toBeCalledWith(
                expect.objectContaining({
                    accessToken: 'azerty',
                    oldPassword: 'My0ld_Pa55w0rD',
                    password: 'Wond3rFu11_Pa55w0rD*$',
                })
            );

            expect(onSuccess).toBeCalled();
            expect(onError).not.toBeCalled();
        });

        test('api update password failed', async () => {
            const user = userEvent.setup();

            updatePassword.mockRejectedValue('Unexpected error');

            await generateComponent({});

            const newPasswordInput = screen.getByLabelText('newPassword');
            expect(newPasswordInput).toBeInTheDocument();

            const passwordConfirmationInput = screen.getByLabelText('passwordConfirmation');
            expect(passwordConfirmationInput).toBeInTheDocument();

            await userEvent.clear(newPasswordInput);
            await userEvent.type(newPasswordInput, 'Wond3rFu11_Pa55w0rD*$');

            await userEvent.clear(passwordConfirmationInput);
            await userEvent.type(passwordConfirmationInput, 'Wond3rFu11_Pa55w0rD*$');

            const submitBtn = screen.getByRole('button', { name: 'send' });
            expect(submitBtn).toBeInTheDocument();

            await user.click(submitBtn);

            expect(onSuccess).not.toBeCalled();
            expect(onError).toBeCalledWith('Unexpected error');
        });
    });
});
