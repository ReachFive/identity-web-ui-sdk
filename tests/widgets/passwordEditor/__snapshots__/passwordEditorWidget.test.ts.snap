// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot password editor basic 1`] = `
.c0 {
  font-size: 14px;
  -webkit-transition: -webkit-transform 400ms ease, opacity 400ms ease;
  -webkit-transition: transform 400ms ease, opacity 400ms ease;
  transition: transform 400ms ease, opacity 400ms ease;
  opacity: 0;
  padding: 20px;
  border-radius: 3px;
  background-color: #ffffff;
  max-width: 400px;
  box-sizing: border-box;
  margin: 0 auto;
}

.c1 {
  --color-primary: #229955;
  --color-destructive: #dc4e41;
  --color-background: #ffffff;
  --color-text: #495057;
  --color-border: #ced4da;
  --spacing-padding-y: 9;
  --spacing-padding-x: 12;
  --spacing-block-inner-height: 38;
  --spacing: 10;
  --font-generic: 14;
  --border-width: 1;
  --radius: 3;
}

.c6 {
  display: block;
  width: 100%;
  box-sizing: border-box;
  text-align: center;
  font-weight: bold;
  vertical-align: middle;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  touch-action: manipulation;
  cursor: pointer;
  color: #ffffff;
  background-color: #229955;
  border: 1px solid #229955;
  padding: 9px 12px;
  font-size: 14px;
  line-height: 1.428571429;
  border-radius: 3px;
  -webkit-transition: all .15s ease-in-out;
  transition: all .15s ease-in-out;
}

.c6:focus {
  outline: 0;
  box-shadow: 0 0 0 3px rgba(34,153,85,0.5);
}

.c6:hover,
.c6:active {
  background-color: #1b7842;
  border-color: #1b7842;
}

.c6[disabled] {
  opacity: .65;
}

.c2 {
  position: relative;
}

.c4 {
  color: #495057;
  margin-bottom: 5px;
  display: none;
}

.c3 {
  margin-bottom: 10px;
}

.c3 > label::after {
  content: "\\A0*";
  color: #dc4e41;
}

.c5 {
  display: block;
  width: 100%;
  box-sizing: border-box;
  font-size: 14px;
  line-height: 1.428571429;
  color: #495057;
  border-radius: 3px;
  border-width: 1px;
  border-style: solid;
  border-color: #ced4da;
  background-color: #fff;
  background-image: none;
  padding: 9px 12px;
  -webkit-transition: border-color ease-in-out .15s, box-shadow ease-in-out .15s;
  transition: border-color ease-in-out .15s, box-shadow ease-in-out .15s;
  box-shadow: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c5:focus {
  border-color: #5fdb94;
  box-shadow: 0 0 0 3px rgba(95,219,148,0.5);
  outline: 0;
}

.c5:disabled,
.c5[readonly] {
  background-color: #e9ecef;
}

.c5::-webkit-input-placeholder {
  color: #868e96;
}

.c5::-moz-placeholder {
  color: #868e96;
}

.c5:-ms-input-placeholder {
  color: #868e96;
}

.c5::placeholder {
  color: #868e96;
}

<div>
  <div
    class="c0 c1 r5-widget"
    transition="entering"
  >
    <form
      class="c2"
      novalidate=""
    >
      <div
        class="c3"
        required=""
      >
        <label
          class="c4"
          for="password"
        >
          newPassword
        </label>
        <div
          style="position: relative;"
        >
          <input
            autocomplete="new-password"
            class="c5"
            data-testid="password"
            id="password"
            name="password"
            placeholder="newPassword"
            required=""
            title="newPassword"
            type="password"
            value=""
          />
        </div>
      </div>
      <div
        class="c3"
        required=""
      >
        <label
          class="c4"
          for="password_confirmation"
        >
          passwordConfirmation
        </label>
        <div
          style="position: relative;"
        >
          <input
            autocomplete="new-password"
            class="c5"
            data-testid="password_confirmation"
            id="password_confirmation"
            name="password_confirmation"
            placeholder="passwordConfirmation"
            required=""
            title="passwordConfirmation"
            type="password"
            value=""
          />
        </div>
      </div>
      <button
        class="c6"
        data-testid="submit"
        type="submit"
      >
        send
      </button>
    </form>
  </div>
</div>
`;
